File: java/1011.java
class Solution {
    public int shipWithinDays(int[] weights, int D) {
        int lo = 0,hi = 0;
        for(int i:weights){
            lo = Math.max(lo,i);
            hi += i;
        }
        
        while(lo < hi){
            int mid = (lo+hi)/2;
            if(helper(mid,weights,D)){
                hi = mid;
            }else{
                lo = mid+1;
            }
        }
        return lo;
    }
    
    public boolean helper(int output,int[] weights,int D){
        int cap = output;
        int size = 1;
        for(int i = 0;i < weights.length;i++){
            if(cap - weights[i] >= 0){
                cap -= weights[i];
            }else{
                if(output < weights[i]){
                    return false;
                }else{
                    cap  = output - weights[i];
                    size ++;
                }
                
            }
        }
        return size <= D;
    }
}

File: python/1011.py
class Solution:
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        
        def validCap(n,D,weights):
            # print(n)
            m = 1
            cap = 0
            for i in range(len(weights)):
                if cap + weights[i] > n:
                    m += 1
                    cap = weights[i]
                else:
                    cap += weights[i]
                
            print(n,m)
            return m <= D
        lo = max(weights)
        hi = sum(weights)
        while lo < hi:
            mid = (lo+hi)//2
            if validCap(mid,D,weights):
                hi = mid
            else:
                lo = mid + 1
        return lo